sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala

```diff
@@ -40,36 +41,47 @@ class CertificateSubmitter
 
   type View = CurrentView[SidechainHistory, SidechainState, SidechainWallet, SidechainMemoryPool]
   type SubmitMessageType = GetDataFromCurrentView[SidechainHistory, SidechainState, SidechainWallet, SidechainMemoryPool, SubmitResult]
+  type CheckMessageType = GetDataFromCurrentView[SidechainHistory, SidechainState, SidechainWallet, SidechainMemoryPool, Unit]
 
   val timeoutDuration: FiniteDuration = settings.scorexSettings.restApi.timeout
   implicit val timeout: Timeout = Timeout(timeoutDuration)
 
+  val prePreviousEpochPlaceholderForFirstEpoch: Array[Byte] = null
+  val missedPrivateKeySignatureBytesPlaceholder:  Array[Byte] = null
+
   override def preStart(): Unit = {
     context.system.eventStream.subscribe(self, classOf[SemanticallySuccessfulModifier[SidechainBlock]])
+    context.system.eventStream.subscribe(self, SidechainAppEvents.SidechainApplicationStart.getClass)
   }
 
   lazy val mainchainApi = new RpcMainchainApi(settings)
 
   protected def submitCertificate(sidechainNodeView: View): SubmitResult = {
     try {
-
+      val history = sidechainNodeView.history
```

Comment:
proof creation and more in general certificate submission should not lock the whole node. Let's execute it in a separate thread independent from nodeviewholder